bigbluebutton-html5/imports/ui/components/whiteboard/whiteboard-overlay/shape-draw-listener/component.jsx

```diff
@@ -0,0 +1,267 @@
+import React, { Component } from 'react';
+import PropTypes from 'prop-types';
+import styles from '../styles.scss';
+
+const MESSAGE_INTERVAL = 50;
+
+export default class ShapeDrawListener extends Component {
+  constructor(props) {
+    super(props);
+
+    this.initialCoordinate = {
+      x: undefined,
+      y: undefined,
+    };
+    this.lastSentCoordinate = {
+      x: undefined,
+      y: undefined,
+    };
+    this.currentCoordinate = {
+      x: undefined,
+      y: undefined,
+    };
+
+    // to track the status of drawing
+    this.isDrawing = false;
+
+    this.currentStatus = undefined;
+
+    // id of the setInterval()
+    this.intervalId = 0;
+
+    this.mouseDownHandler = this.mouseDownHandler.bind(this);
+    this.mouseMoveHandler = this.mouseMoveHandler.bind(this);
+    this.mouseUpHandler = this.mouseUpHandler.bind(this);
+    this.resetState = this.resetState.bind(this);
+    this.sendLastMessage = this.sendLastMessage.bind(this);
+    this.sendCoordinates = this.sendCoordinates.bind(this);
+  }
+
+  componentDidMount() {
+    // to send the last message if the user refreshes the page while drawing
+    window.addEventListener('beforeunload', this.sendLastMessage);
+  }
+
+  componentWillUnmount() {
+    window.removeEventListener('beforeunload', this.sendLastMessage);
+
+    // sending the last message on componentDidUnmount
+    this.sendLastMessage();
+  }
+
+  // main mouse down handler
+  mouseDownHandler(event) {
+    // Sometimes when you Alt+Tab while drawing it can happen that your mouse is up,
+    // but the browser didn't catch it. So check it here.
+    if (this.isDrawing) {
+      return this.sendLastMessage();
+    }
+
+    window.addEventListener('mouseup', this.mouseUpHandler);
+    window.addEventListener('mousemove', this.mouseMoveHandler, true);
+    this.isDrawing = true;
+
+    const {
+      getTransformedSvgPoint,
+      generateNewShapeId,
+      svgCoordinateToPercentages,
+    } = this.props.actions;
+
+    // sending the first message
+    let transformedSvgPoint = getTransformedSvgPoint(event);
+
+    // transforming svg coordinate to percentages relative to the slide width/height
+    transformedSvgPoint = svgCoordinateToPercentages(transformedSvgPoint);
+
+    // generating new shape id
+    generateNewShapeId();
```

Comment:
`id` is controlled by the parent, WhiteboardOverlay.
It is [saved](https://github.com/OZhurbenko/bigbluebutton/blob/f3a17e74a834231b094d159dc4c260968a16f5d6/bigbluebutton-html5/imports/ui/components/whiteboard/whiteboard-overlay/component.jsx#L30) and [generated](https://github.com/OZhurbenko/bigbluebutton/blob/f3a17e74a834231b094d159dc4c260968a16f5d6/bigbluebutton-html5/imports/ui/components/whiteboard/whiteboard-overlay/component.jsx#L79) there, and then each handler has 2 functions to call - `generateNewShapeId` and `getCurrentShapeId`.

Another way to do it would be to save `id` locally in the draw listeners, I just thought it would be easier to read if it's saved where it was created.